src/plugins/files/server/blob_storage_service/adapters/es/es.ts

```diff
@@ -187,7 +206,26 @@ export class ElasticsearchBlobStorageClient implements BlobStorageClient {
     // right after uploading it, we refresh the index before downloading the file.
     await this.esClient.indices.refresh({ index: this.index });
 
-    return this.getReadableContentStream(id, size);
+    const downloadRequest$ = defer(this.getReadableContentStream(id, size)).pipe(
+      this.downloadSemaphore.acquire()
+    );
+
+    return new Readable({
+      read(this, rsSize) {
+        downloadRequest$
+          .pipe(bufferCount(rsSize))
+          .pipe(
+            tap((buffer) => {
+              this.push(Buffer.concat(buffer));
+            })
+          )
+          .subscribe({
+            complete: () => {
+              this.push(null);
+            },
+          });
+      },
+    });
```

Comment:
Could it be done without introducing `Readable` implementation here? Somehow like this:

<img width="1199" alt="image" src="https://github.com/elastic/kibana/assets/82822460/64537a93-c2cc-4069-a781-e47900b9cc74">

Reply:
in it's current form both operators `firstValueFrom` and `lastValueFrom` yield incomplete chunks of the document requested, I've simplified this so that `firstValueFrom` actual resolves to the entire stream, hence no need for introducing the Readable implementation